src/main/java/org/opensearch/flowframework/rest/AbstractSearchWorkflowAction.java

```diff
@@ -0,0 +1,98 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.rest;
+
+import org.opensearch.action.ActionType;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.rest.BaseRestHandler;
+import org.opensearch.rest.BytesRestResponse;
+import org.opensearch.rest.RestChannel;
+import org.opensearch.rest.RestRequest;
+import org.opensearch.rest.RestResponse;
+import org.opensearch.rest.action.RestResponseListener;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.opensearch.core.xcontent.ToXContent.EMPTY_PARAMS;
+import static org.opensearch.flowframework.util.RestHandlerUtils.getSourceContext;
+
+/**
+ * Abstract class to handle search request.
+ */
+public class AbstractSearchWorkflowAction<T extends ToXContentObject> extends BaseRestHandler {
+
+    protected final List<String> urlPaths;
+    protected final String index;
+    protected final Class<T> clazz;
+    protected final ActionType<SearchResponse> actionType;
+
+    /**
+     * Instantiates a new AbstractSearchWorkflowAction
+     * @param urlPaths urlPaths to create routes
+     * @param index index the search should be done on
+     * @param clazz model class
+     * @param actionType from which action abstract class is called
+     */
+    public AbstractSearchWorkflowAction(List<String> urlPaths, String index, Class<T> clazz, ActionType<SearchResponse> actionType) {
+        this.urlPaths = urlPaths;
+        this.index = index;
+        this.clazz = clazz;
+        this.actionType = actionType;
+    }
+
+    @Override
+    public String getName() {
+        return null;
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.parseXContent(request.contentOrSourceParamParser());
+        searchSourceBuilder.fetchSource(getSourceContext(request, searchSourceBuilder));
+        searchSourceBuilder.seqNoAndPrimaryTerm(true).version(true);
+        SearchRequest searchRequest = new SearchRequest().source(searchSourceBuilder).indices(index);
+        return channel -> client.execute(actionType, searchRequest, search(channel));
+    }
+
+    /**
+     * Builds the action response for the Search Request
+     *
+     * @param channel the REST channel
+     * @return the action response
+     */
+    protected RestResponseListener<SearchResponse> search(RestChannel channel) {
```

Comment:
Let's handle the `onFailure` case here and wrap any exception with a `FlowFrameworkException`. I suggest rebasing with PR https://github.com/opensearch-project/opensearch-ai-flow-framework/pull/137 after its merged in. It modifies the `FlowFrameworkException` to implement `ToXContentObject` so that exceptions can be outputted in the correct format as part of a rest response.

Reply:
Here the error is read from `SearchResponse` itself and not from the `exception`
```
 return new BytesRestResponse(RestStatus.REQUEST_TIMEOUT, response.toString());
 ```